drivers/i2s/i2s_nrfx.c

```diff
@@ -0,0 +1,1309 @@
+		interface_error_service();
+		channel_error_service(ch_tx);
+		return;
+	}
+	/* continue transmission */
+	p_new_buffers->p_tx_buffer = mem_block;
+}
+
+static void channel_rx_mem_clear(void const *first_block)
+{
+	struct i2s_nrfx_interface *i2s = get_interface();
+	struct channel_str * const ch_rx = i2s->channel_rx;
+	void *mem_block;
+	size_t mem_block_size;
+
+	while (queue_fetch(&ch_rx->mem_block_queue,
+			(void **)&mem_block,
+			&mem_block_size) != 0) {
+		if (k_sem_take(&ch_rx->sem, ch_rx->timeout * 10) < 0) {
+			return;
+		}
+		k_mem_slab_free(ch_rx->mem_slab,
+			(void **)&mem_block);
+	}
+}
+
+static void channel_rx_callback(nrfx_i2s_buffers_t const *p_released,
+				u32_t status, nrfx_i2s_buffers_t *p_new_buffers)
+{
+	struct i2s_nrfx_interface *i2s = get_interface();
+	struct channel_str *ch_rx = i2s->channel_rx;
+	int ret;
+
+	if (p_released->p_rx_buffer != NULL && next_buffers_needed(status)) {
+		ret = ch_rx->get_data((u32_t **)&p_released->p_rx_buffer,
+				      &i2s->size);
+		if (ret < 0) {
+			return;
+		}
+	}
+	if (ch_rx->current_state == I2S_STATE_STOPPING) {
+		enum i2s_trigger_cmd ch_cmd = ch_rx->last_trigger_cmd;
+
+		if (next_buffers_needed(status)) {
+			switch (ch_cmd) {
+			case I2S_TRIGGER_DROP:
+				ch_rx->mem_clear(NULL);
+				break;
+			case I2S_TRIGGER_DRAIN:
+			case I2S_TRIGGER_STOP:
+				break;
+			default:
+				channel_error_service(ch_rx);
+				return;
+			}
+		} else {
+			if (p_released->p_rx_buffer) {
+				ret = ch_rx->get_data(
+					(u32_t **)&p_released->p_rx_buffer,
+					&i2s->size);
+				if (ret < 0) {
+					return;
+				}
+			}
+			int ret = channel_change_state(ch_rx,
+					I2S_STATE_READY);
+			if (ret < 0) {
+				channel_error_service(ch_rx);
+				return;
+			}
+		}
+		i2s->buffers.p_rx_buffer = NULL;
+		return;
+	} else if (ch_rx->current_state == I2S_STATE_RUNNING &&
+		 i2s->get_state() == I2S_IF_NEEDS_RESTART) {
+		return;
+	} else if (ch_rx->current_state == I2S_STATE_ERROR) {
+		return;
+	}
+	if (next_buffers_needed(status)) {
+		ret = k_mem_slab_alloc(ch_rx->mem_slab,
+				       (void **)&p_new_buffers->p_rx_buffer,
+				       K_NO_WAIT);
+		if (ret < 0) {
+			/*overrun error occurred*/
+			LOG_ERROR;
+			interface_error_service();
+			channel_error_service(ch_rx);
+			return;
+		}
+	}
+}
+
+static int channel_rx_get_data(u32_t **buf, size_t *block_size)
+{
+	struct channel_str *ch_rx = get_interface()->channel_rx;
+	int ret = queue_add(&ch_rx->mem_block_queue, *buf, *block_size);
+
+	if (ret < 0) {
+		return ret;
+	}
+	k_sem_give(&ch_rx->sem);
+	return 0;
+}
+
+static void isr(void *arg)
+{
+	struct i2s_nrfx_interface *i2s = get_interface();
+
+	nrfx_i2s_irq_handler();
+	if (i2s->get_state() == I2S_IF_RESTARTING) {
+		int ret = i2s->start();
+
+		if (ret < 0) {
+			i2s->error_service();
+		}
+		return;
+	}
+}
+
+
+#ifdef CONFIG_NRFX_I2S
+struct queue_item rx_1_ring_buf[CONFIG_NRFX_I2S_RX_BLOCK_COUNT + 1];
+struct queue_item tx_1_ring_buf[CONFIG_NRFX_I2S_TX_BLOCK_COUNT + 1];
```

Comment:
These arrays should be `static`. BTW what's the meaning of this `1` in their names?